utils/src/Vulkan/Utils/Reflect.hs

```diff
@@ -0,0 +1,567 @@
+  { stage   = shaderStageFlagBits stage
+  , module' = shaderModule
+  , name    = B.pack . T.unpack $ name
+  }
+
+makePipelineShaderStageCreateInfos
+  :: ShaderStage
+  -> Vector EntryPoint
+  -> ShaderModule
+  -> Vector (PipelineShaderStageCreateInfo '[])
+makePipelineShaderStageCreateInfos stage entryPoints shaderModule =
+  V.map (makePipelineShaderStageCreateInfo stage shaderModule) entryPoints
+
+-- https://github.com/KhronosGroup/GLSL/blob/master/extensions/khr/GL_KHR_vulkan_glsl.txt
+-- https://vulkan.lunarg.com/doc/view/1.2.135.0/linux/tutorial/html/08-init_pipeline_layout.html
+makeUboDescriptorSetLayoutBinding
+  :: ShaderStage -> Ubo -> (Int, DescriptorSetLayoutBinding)
+makeUboDescriptorSetLayoutBinding stage Ubo {..} =
+  ( set
+  , zero { binding         = fromIntegral binding
+         , descriptorType  = DESCRIPTOR_TYPE_UNIFORM_BUFFER
+         , descriptorCount = maybe 1 (fromIntegral . V.sum) array
+         , stageFlags      = shaderStageFlagBits stage
+         }
+  )
+
+data TextureDescriptorType = Sampler2D
+  deriving Show
+
+textureDescriptorTypeFromText :: Text -> Maybe TextureDescriptorType
+textureDescriptorTypeFromText = \case
+  "sampler2D" -> Just Sampler2D
+  _           -> Nothing
+
+textureDescriptorTypeToText :: TextureDescriptorType -> Text
+textureDescriptorTypeToText = \case
+  Sampler2D -> "sampler2D"
+
+instance FromJSON TextureDescriptorType where
+  parseJSON = withTextMaybe "type" textureDescriptorTypeFromText
+
+instance ToJSON TextureDescriptorType where
+  toJSON = String . textureDescriptorTypeToText
+
+convertTextureDescriptorType :: TextureDescriptorType -> DescriptorType
+convertTextureDescriptorType = \case
+  Sampler2D -> DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER
+
+makeTextureDescriptorSetLayoutBinding
+  :: ShaderStage -> Texture -> (Int, DescriptorSetLayoutBinding)
+makeTextureDescriptorSetLayoutBinding stage Texture {..} =
+  ( set
+  , zero { binding         = fromIntegral binding
+         , descriptorType  = convertTextureDescriptorType type'
+         , descriptorCount = maybe 1 (fromIntegral . V.sum) array
+         , stageFlags      = shaderStageFlagBits stage
+         }
+  )
+
+makeDescriptorSetLayoutBindings
+  :: ShaderStage
+  -> Vector Ubo
+  -> Vector Texture
+  -> Vector (Int, DescriptorSetLayoutBinding)
+makeDescriptorSetLayoutBindings stage ubos textures =
+  uboBindings <> textureBindings
+ where
+  uboBindings = V.map (makeUboDescriptorSetLayoutBinding stage) ubos
+  textureBindings =
+    V.map (makeTextureDescriptorSetLayoutBinding stage) textures
+
+makeDescriptorSetLayoutCreateInfos
+  :: Vector (Int, DescriptorSetLayoutBinding)
+  -> V.Vector (DescriptorSetLayoutCreateInfo '[])
+makeDescriptorSetLayoutCreateInfos bindings =
+  V.map makeDescriptorSetLayoutCreateInfo
+    . V.fromList
+    . M.elems
+    . M.unionWith (V.++) sets
+    $ setsMap
+ where
+  setSize = V.maximum . (fst <$>) $ bindings :: Int
```

Comment:
`V.maximumBy fst bindings`

Reply:
`V.maximum . (fst <$>) :: Ord c => Vector (c, b) -> c`

`V.maximumBy :: (a -> a -> Ordering) -> Vector a -> a`

I changed it to `V.maximum . fmap fst $ bindings`